include/flang/Optimizer/Dialect/FIROps.td

```diff
@@ -0,0 +1,2779 @@
+    pass. An undefined value can be of any type except `!fir.ref<T>`.
+
+      %a = fir.undefined !fir.array<10 x !fir.type<T>>
+
+    The example creates an array shaped ssa value. The array is rank 1, extent
+    10, and each element has type `!fir.type<T>`.
+  }];
+
+  let verifier = [{
+    if (auto ref = getType().dyn_cast<fir::ReferenceType>())
+      return emitOpError("undefined values of type !fir.ref not allowed");
+    return mlir::success();
+  }];
+
+  let parser = [{
+    mlir::Type intype;
+    if (parser.parseType(intype) ||
+        parser.parseOptionalAttrDict(result.attributes) ||
+        parser.addTypeToList(intype, result.types))
+      return mlir::failure();
+    return mlir::success();
+  }];
+
+  let printer = [{
+    p << getOperationName() << ' ' << getType();
+    p.printOptionalAttrDict(getAttrs(), {});
+  }];
+}
+
+def fir_AllocMemOp : fir_AllocatableOp<"allocmem"> {
+  let summary = "allocate storage on the heap for an object of a given type";
+
+  let description = [{
+    Creates a heap memory reference suitable for storing a value of the
+    given type, T.  The heap refernce returned has type `!fir.heap<T>`.
+    The memory object is in an undefined state.  `allocmem` operations must
+    be paired with `freemem` operations to avoid memory leaks.
+
+      %0 = fir.allocmem !fir.array<10 x f32>
+      fir.freemem %0 : !fir.heap<!fir.array<10 x f32>>
+  }];
+
+  let results = (outs fir_HeapType);
+
+  let verifier = allocVerify#[{
+    mlir::Type outType = getType();
+    if (!outType.dyn_cast<fir::HeapType>())
+      return emitOpError("must be a !fir.heap type");
+    return mlir::success();
+  }];
+
+  let extraClassDeclaration = extraAllocClassDeclaration#[{
+    static mlir::Type wrapResultType(mlir::Type intype);
+  }];
+}
+
+def fir_FreeMemOp : fir_Op<"freemem", []> {
+  let summary = "free a heap object";
+
+  let description = [{
+    Deallocates a heap memory reference that was allocated by an `allocmem`.
+    The memory object that is deallocated is placed in an undefined state
+    after `fir.freemem`.  Optimizations may treat the loading of an object
+    in the undefined state as undefined behavior.  This includes aliasing
+    references, such as the result of an `fir.embox`.
+
+      %21 = fir.allocmem !fir.type<ZT(p:i32){field:i32}>
+      ...
+      fir.freemem %21 : !fir.heap<!fir.type<ZT>>
+  }];
+
+  let arguments = (ins fir_HeapType:$heapref);
+
+  let assemblyFormat = "$heapref attr-dict `:` type($heapref)";
+}
+
+// Terminator operations
+
+class fir_SwitchTerminatorOp<string mnemonic, list<OpTrait> traits = []> :
+    fir_Op<mnemonic, !listconcat(traits, [Terminator])> {
+  let arguments = (ins Variadic<AnyType>:$args);
+
+  let results = (outs);
+
+  let builders = [OpBuilder<
+    "Builder *, OperationState &result, Value selector,"
+    "ArrayRef<Value> properOperands, ArrayRef<Block *> destinations,"
+    "ArrayRef<ArrayRef<Value>> operands = {},"
```

Comment:
Generally `ValueRange` is a better parameter than `ArrayRef<Value>` to use for builders, as it can be constructed from operand/result/block/etc. ranges without constructing a vector.